### PR TITLE
Ldpc v2

### PR DESCRIPTION
**Pull request by quantumgizmos**
Complete rewrite. Should be backwards compatible with LDPCv1. Please test this on old scripts.

### CODE REVIEW

File: cpp_example/main.cpp [hunk 1]
```diff
@@ -0,0 +1,13 @@
+#include <iostream>
```
Comment (lucasberent) [inline cpp_example/main.cpp:1]: is this file needed?

File: cpp_test/TestFlipDecoder.cpp [hunk 2]
```diff
@@ -0,0 +1,121 @@
...
+    pcm.insert_entry(3, 0), pcm.insert_entry(3, 1), pcm.insert_entry(3, 2), pcm.insert_entry(3, 7);
+
+    ldpc::sparse_matrix_util::print_sparse_matrix(pcm);
+
+    auto flipD = new ldpc::flip::FlipDecoder(pcm, 8);
+
+
+    cout<<endl;
+
+    for(int i = 2; i<8; i++){
```
Comment (lucasberent) [inline cpp_test/TestFlipDecoder.cpp:23]: I believe for indexing `std::size_t` is best to use as loop variable type

File: cpp_test/TestFlipDecoder.cpp [hunk 3]
```diff
@@ -0,0 +1,121 @@
...
+
+        auto error = vector<uint8_t>(pcm.n,0);
+        error[i] = 1;
+
+        vector<uint8_t> syndrome(pcm.m, 0);
+        pcm.mulvec(error,syndrome);
+
+        auto decoding = flipD->decode(syndrome);
+
+        cout<<"Error: ";
```
Comment (lucasberent) [inline cpp_test/TestFlipDecoder.cpp:33]: should we leave in all cout statements in the release?

File: cpp_test/TestGf2Dense.cpp [hunk 4]
```diff
@@ -0,0 +1,359 @@
...
+        csr_matrix.push_back(row_indices);  // Store the column indices of non-zero elements
+        row_ptr.push_back(row_ptr.back() + nnz);  // Update row_ptr
+    }
+    
+    // Optionally, you can also return the row_ptr
+    // For this example, I'm only returning the csr_matrix containing column indices of non-zero elements
+    return csr_matrix;
+}
+
+// TEST(gf2dense, hamming_code){
```
Comment (lucasberent) [inline cpp_test/TestGf2Dense.cpp:49]: what about all this outcommented code?

File: cpp_test/TestGf2Linalg.cpp [hunk 5]
```diff
@@ -0,0 +1,308 @@
...
+
+    auto ker0 = rnd_mat.matmul(kerT);
+
+    auto expected = GF2Sparse<>(m,n);
+
+    ASSERT_TRUE(ker0==expected);
+
+}
+
+double sparsity = 0.01;
```
Comment (lucasberent) [inline cpp_test/TestGf2Linalg.cpp:262]: unsure if blobal variables adhere to the best practices here.

File: cpp_test/TestSoftInfo.cpp [hunk 6]
```diff
@@ -0,0 +1,183 @@
+#include <gtest/gtest.h>
+#include "bp.hpp"
+#include "ldpc.hpp"
+
+#include "osd.hpp"
+
+using namespace std;
+
```
Comment (lucasberent) [inline cpp_test/TestSoftInfo.cpp:8]: lots of code commented out here

File: cpp_test/TestUnionFind.cpp [hunk 7]
```diff
@@ -0,0 +1,130 @@
...
+#include "gf2codes.hpp"
+#include "union_find.hpp"
+#include "util.hpp"
+#include "bp.hpp"
+
+using namespace std;
+using namespace ldpc::uf;
+
+// TEST(UfDecoder, single_bit_error) {
+
```
Comment (lucasberent) [inline cpp_test/TestUnionFind.cpp:12]: commented out code

File: cpp_test/test_inputs/generate_tests.py [hunk 8]
```diff
@@ -0,0 +1,220 @@
...
+        output_vector = vector_to_string(output_vector)
+
+        csv_string = ""
+        m, n = pcm.shape
+        csr = to_csr(pcm)
+        csv_string += f"{m};{n};{csr};{output_vector}"
+        print(csv_string, file=output_file)
+
+
+if __name__ == "__main__":
```
Comment (lucasberent) [inline cpp_test/test_inputs/generate_tests.py:218]: is a main function needed here?

File: python_test/test_bp_decoder.py [hunk 9]
```diff
@@ -0,0 +1,250 @@
...
+    mc2 = MonteCarloBscSimulation(H, error_rate=error_rate, Decoder=bpd_v1, target_run_count=run_count,seed=42)
+    out1 = mc1.run()
+    out2 = mc2.run()
+
+
+    assert out1['logical_error_rate'] == out2['logical_error_rate']
+
+
+
+if __name__ == "__main__":
```
Comment (lucasberent) [inline python_test/test_bp_decoder.py:249]: needed?

File: python_test/test_code_util.py [hunk 10]
```diff
@@ -0,0 +1,81 @@
...
+import scipy.sparse
+import numpy as np
+from ldpc import mod2
+from ldpc import code_util
+
+def test_construct_generator_matrix():
+
+    for i in range(3,10):
+
+        H = hamming_code(i)
```
Comment (lucasberent) [inline python_test/test_code_util.py:11]: I guess using pytest.fixture would be the best practice to generate test objects used (also those used across different tests)

**Review by lucasberent — APPROVED** [review]
Checked the most important parts hopefully and added a few comments. I guess more hands on testing would be valuable.

**Review by lucasberent — CHANGES_REQUESTED** [review]
some minor comments and potential improvements on the sinter integration.

File: src_python/ldpc/sinter_decoders/sinter_bposd_decoder.py [hunk 11]
```diff
@@ -29,23 +63,33 @@ def decode_via_files(
     ) -> None:
         self.dem = stim.DetectorErrorModel.from_file(dem_path)
         self.matrices = detector_error_model_to_check_matrices(self.dem)
-        self.bposd = BpOsdDecoder(self.matrices.check_matrix, error_channel=list(self.matrices.priors), max_iter=5, bp_method="ms", ms_scaling_factor = 0.625, schedule="parallel", osd_method = "osd0")
+        self.bposd = BpOsdDecoder(self.matrices.check_matrix,
```
[the remark this answers is not in the document]
Comment (peter-janderks) [reply src_python/ldpc/sinter_decoders/sinter_bposd_decoder.py:99]: If I don't make a separate sinter_[decoder_name]_ for every decoder, then someone using Sinter in python can't see the the parameters the sinter_[decoder_name]_ excepts in their editor.  Do you know what I mean? For a Sinter user (who I'm targeting) I think this solution is the cleanest.

File: src_python/ldpc/sinter_decoders/sinter_bposd_decoder.py [hunk 12]
```diff
@@ -29,23 +63,33 @@ def decode_via_files(
...
+
         stim.write_shot_data_file(
             data=predictions,
             path=obs_predictions_b8_out_path,
             format="b8",
             num_observables=self.dem.num_observables,
         )
 
-
     def decode(self, syndrome: np.ndarray) -> np.ndarray:
```
[the remark this answers is not in the document]
Comment (peter-janderks) [reply src_python/ldpc/sinter_decoders/sinter_bposd_decoder.py:125]: Yes agreed, I made an example in [sinter_example.py](https://github.com/quantumgizmos/ldpc/blob/ldpc_v2/examples/sinter_example.py). Can you let me know what you think? To be clear, I don't want to write code that replaces Sinter and Stim, I simply want to use the BPOSD and BeliefFind as a plugin. Sinter and Stim have a full simulator that only takes a code and a decoder/-name.

File: src_python/ldpc/sinter_decoders/sinter_bposd_decoder.py [hunk 13]
```diff
@@ -1,20 +1,54 @@
 import stim
 import numpy as np
 import pathlib
-from beliefmatching import detector_error_model_to_check_matrices
 from ldpc.bposd_decoder import BpOsdDecoder
 from sinter import Decoder
+from beliefmatching import detector_error_model_to_check_matrices
```
[the remark this answers is not in the document]
Comment (peter-janderks) [reply src_python/ldpc/sinter_decoders/sinter_bposd_decoder.py:6]: I don't want to copy Oscar's code from beliefmatching. So I'll leave it as it is now. (Sidenote- if we want to optimize performance I don't think we should not convert detector error models to matrices).

File: src_python/ldpc/sinter_decoders/sinter_bposd_decoder.py [hunk 14]
```diff
@@ -1,20 +1,54 @@
...
         self.ms_scaling_factor = ms_scaling_factor
         self.schedule = schedule
+        self.omp_thread_count = omp_thread_count
+        self.random_serial_schedule = random_serial_schedule
+        self.serial_schedule_order = serial_schedule_order
         self.osd_method = osd_method
+        self.osd_order = osd_order
 
     def decode_via_files(
         self,
```
[the remark this answers is not in the document]
Comment (peter-janderks) [reply src_python/ldpc/sinter_decoders/sinter_bposd_decoder.py:58]: Fixed now. The tmp_dir parameter is not optional

**Comment by peter-janderks** [comment]
@lucasberent Thanks for the quick review 😃 . There are two comments left which I didn't resolve yet but I did comment on.

File: src_python/ldpc/sinter_decoders/sinter_bposd_decoder.py (hunk 11 above)
[the remark this answers is not in the document]
Comment (lucasberent) [reply src_python/ldpc/sinter_decoders/sinter_bposd_decoder.py:99]: If you used named arguments that should work, no?

**Comment by lucasberent** [comment]
> @lucasberent Thanks for the quick review 😃 . There are two comments left which I didn't resolve yet but I did comment on.

haven't tried it out myself yet but looks good! thanks

**Pull request merged**